application/src/main/java/org/togetherjava/tjbot/db/Database.java

```diff
@@ -0,0 +1,187 @@
+package org.togetherjava.tjbot.db;
+
+
+import org.flywaydb.core.Flyway;
+import org.jooq.DSLContext;
+import org.jooq.SQLDialect;
+import org.jooq.exception.DataAccessException;
+import org.jooq.impl.DSL;
+import org.sqlite.SQLiteConfig;
+import org.sqlite.SQLiteConfig.JournalMode;
+import org.sqlite.SQLiteDataSource;
+import org.togetherjava.tjbot.util.CheckedConsumer;
+import org.togetherjava.tjbot.util.CheckedFunction;
+
+import java.sql.SQLException;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * The main database class.
+ */
+public class Database {
+
+    private final DSLContext dslContext;
+    private final Lock writeLock = new ReentrantLock();
+
+    /**
+     * Creates a new database.
+     *
+     * @param jdbcUrl the url to the database
+     * @throws SQLException if no connection could be established
+     */
+    public Database(String jdbcUrl) throws SQLException {
+        SQLiteConfig sqliteConfig = new SQLiteConfig();
+        sqliteConfig.enforceForeignKeys(true);
+        // In WAL mode only concurrent writes pose a problem so we synchronize on those
+        sqliteConfig.setJournalMode(JournalMode.WAL);
+
+        SQLiteDataSource dataSource = new SQLiteDataSource(sqliteConfig);
+        dataSource.setUrl(jdbcUrl);
+
+        Flyway flyway = Flyway.configure().dataSource(dataSource).locations("/db").load();
+        flyway.migrate();
+
+        this.dslContext = DSL.using(dataSource.getConnection(), SQLDialect.SQLITE);
+    }
+
+    /**
+     * Acquires read access to the database.
+     *
+     * @return a way to interact with the database in a read only way
+     * @throws DatabaseException if an error occurs in the passed handler function
+     */
+    public <T> T read(CheckedFunction<DSLContext, T, DataAccessException> action) {
+        try {
+            return action.accept(getDslContext());
+        } catch (DataAccessException e) {
+            throw new DatabaseException(e);
+        }
+    }
+
+    /**
+     * Acquires read access to the database.
+     *
+     * @throws DatabaseException if an error occurs in the passed handler function
+     */
+    public void read(CheckedConsumer<DSLContext, DataAccessException> action) {
+        read(ctx -> {
+            action.accept(ctx);
+            return null;
+        });
+    }
+
+    /**
+     * Acquires read and write access to the database.
+     *
+     * @return a way to interact with the database
+     * @throws DatabaseException if an error occurs in the passed handler function
+     */
+    public <T> T write(CheckedFunction<DSLContext, T, DataAccessException> action) {
+        try {
+            writeLock.lock();
```

Comment:
This lock prevents two threads from writing to the database concurrently (hence the name `writeLock`). Reading shouldn't be an issue as the WAL *should* protect us against concurrent reads and reads with a *single* concurrent writer.

Reply:
I dont know if sqllite behaves differently but database issue atleast a row lock whenever a row is updated. Hence writes are always atomic per row at the least.

Reply:
Every write should be atomic unless the database is broken. But sqlite doesn't like concurrent writes as it might find a busy database and *fail the operation*. So you would need to retry writes manually, the lock there just prevents it from happening entirely which is a bit easier.

Reply:
Hmm..... sounds good then 👍